src/map/src/combat/combat.cpp

```diff
@@ -215,56 +222,135 @@ void Combat::update(EntitySystem& entitySystem, Entity entity, uint32_t dt) {
+        break;
       } else {
         logger->debug("applied {} damage to entity '{}' {}.", attack.value_, basicInfo.name, basicInfo.id);
-
         auto p = SrvDamage::create(attack.attacker_, basicInfo.id, attack.value_, attack.action_);
         entitySystem.send_to(entity, p);
         entitySystem.send_to(attacker, p);
       }
+
+      // Add this damage to the combat log
+      bool found_log = false;
+      for(auto& attack_log : queuedDamage.damage_log_)
+      {
+        if(attack_log.attacker_ == attack.attacker_)
+        {
+          attack_log.value_ += attack.value_;
+          found_log = true;
+          break;
+        }
+      }
+
+      if(found_log == false)
+      {
+        queuedDamage.damage_log_.push_back(attack);
+      }
     }
     life.hp = adjusted_hp;
     queuedDamage.damage_.clear();
-    
+
     if(life.hp <= 0) {
-      if(entitySystem.has_component<Component::Mob>(entity) == true) {
+      if(entitySystem.has_component<Component::Mob>(entity) == true)
+      {
+        auto& npcLua = entitySystem.get_component<Component::NpcLua>(entity);
+        auto data = npcLua.data.lock();
+        int32_t give_exp = data->get_give_exp();
+        int32_t xp_out = 1;
+        int32_t level_difference = 0;
+        int32_t server_rate = 100; // this is the default for a 1x server
+        for(auto& attack_log : queuedDamage.damage_log_)
+        {
+          Entity attacker = entitySystem.get_entity_from_id(attack_log.attacker_);
+          auto& attackerLevel = entitySystem.get_component<Component::Level>(attacker);
+
+          // Start exp out algo
+          level_difference = attackerLevel.level - level.level;
+          if(attack_log.value_ > (life.maxHp * 1.15f))
+            attack_log.value_ = (life.maxHp * 1.15f);
+
+          float dmgModifier = (attack_log.value_ + life.maxHp / 15.0f + 30);
+          auto expCalc = ((level.level + 3) * give_exp * dmgModifier);
+
+          if(level_difference <= 3)
+            xp_out = expCalc * server_rate / life.maxHp / 370.0f;
+          else if(level_difference < 9)
+            xp_out = expCalc * server_rate / life.maxHp / (level_difference + 3) / 60.0f;
+          else
+            xp_out = expCalc * server_rate / life.maxHp / (level_difference + 3) / 180.0f;
+
+          if(xp_out < 1)
+            xp_out = 1;
+
+          attackerLevel.xp += xp_out;
+          // End exp out algo
+
+          auto exp_required = get_exp_to_level(attackerLevel.level+1);
+          while (attackerLevel.xp >= exp_required) {
+            int32_t new_exp = attackerLevel.xp - exp_required;
+            attackerLevel.xp = new_exp;
+            ++attackerLevel.level;
+            exp_required = get_exp_to_level(attackerLevel.level+1);
+
+            //TODO:: adjust our hp and mp based on our level here
+            //TODO:: give the player attr and skill points
+            auto p = SrvLevelup::create(attack_log.attacker_, attackerLevel.level, attackerLevel.xp, 0, 0);
```

Comment:
maybe have a dedicated function to take care of the whole level up thingy?

Reply:
Will do this in next combat pass